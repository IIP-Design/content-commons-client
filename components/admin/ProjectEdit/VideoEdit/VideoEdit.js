/**
 *
 * VideoEdit
 *
 */
import React, { useState, useEffect } from 'react';
import { withRouter } from 'next/router';
import PropTypes from 'prop-types';

import { graphql } from 'react-apollo';
import compose from 'lodash.flowright';
import { connect } from 'react-redux';
import * as actions from 'lib/redux/actions/upload';

import { buildUpdateVideoProjectTree } from 'lib/graphql/builders/video';
import { buildThumbnailTree } from 'lib/graphql/builders/common';
import {
  UPDATE_VIDEO_PROJECT_MUTATION,
  DELETE_VIDEO_PROJECT_MUTATION,
  UPDATE_VIDEO_UNIT_MUTATION,
  VIDEO_PROJECT_FILES_QUERY,
  VIDEO_PROJECT_UNITS_QUERY,
  VIDEO_PROJECT_QUERY,
} from 'lib/graphql/queries/video';
import { IMAGE_USES_QUERY } from 'lib/graphql/queries/common';

import ActionButtons from 'components/admin/ActionButtons/ActionButtons';
import ApolloError from 'components/errors/ApolloError';
import Notification from 'components/Notification/Notification';
import ProjectHeader from 'components/admin/ProjectHeader/ProjectHeader';
import ProjectUnits from 'components/admin/ProjectUnits/ProjectUnits';
import VideoProjectDetailsForm from 'components/admin/ProjectDetailsForm/VideoProjectDetailsForm/VideoProjectDetailsForm';
import VideoProjectSupportFiles from 'components/admin/ProjectSupportFiles/VideoProjectSupportFiles/VideoProjectSupportFiles';
import UploadProgress from 'components/admin/ProjectEdit/UploadProgress/UploadProgress';
import useTimeout from 'lib/hooks/useTimeout';
import withFileUpload from 'hocs/withFileUpload/withFileUpload';
import { UploadContext } from './UploadContext';
import './VideoEdit.scss';

const VideoEdit = props => {
  const { setIsUploading } = props;

  const MAX_CATEGORY_COUNT = 2;
  const SAVE_MSG_DELAY = 2000;

  const [mounted, setMounted] = useState( false );
  const [error, setError] = useState( {} );
  // eslint-disable-next-line react/destructuring-assignment
  const [projectId, setProjectId] = useState( props.id );
  const [deleteConfirmOpen, setDeleteConfirmOpen] = useState( false );
  const [isFormValid, setIsFormValid] = useState( true );

  const [notification, setNotification] = useState( {
    notificationMessage: '',
    showNotification: false,
  } );

  const updateNotification = msg => {
    setNotification( {
      notificationMessage: msg,
      showNotification: !!msg,
    } );
  };

  const { startTimeout } = useTimeout( () => {
    if ( mounted ) {
      updateNotification( '' );
    }
  }, SAVE_MSG_DELAY );

  const [disableBtns, setDisableBtns] = useState( false );

  useEffect( () => {
    const { videoProjectUnitsQuery } = props;

    if ( videoProjectUnitsQuery ) {
      const { projectUnits } = videoProjectUnitsQuery;

      if ( projectUnits && projectUnits.units.length < 1 ) setDisableBtns( true );
    }
  }, [] );

  const centeredStyles = {
    position: 'absolute',
    top: '1em',
    left: '50%',
    transform: 'translateX(-50%)',
  };

  const uploadVideoFileNotificationStyles = {
    display: 'block',
    fontSize: '1em',
    textAlign: 'center',
    backgroundColor: '#cd2026',
    color: 'white',
    opacity: 0,
  };

  const {
    upload: { isUploading, filesToUpload },
  } = props;

  useEffect( () => {
    setMounted( true );

    return () => {
      setMounted( false );
      props.uploadReset(); // clear files to upload store
    };
  }, [] );

  const addProjectIdToUrl = id => {
    const { router } = props;

    const path = `${router.asPath}&id=${id}`;

    router.replace( router.asPath, path, { shallow: true } );
  };

  const deleteProjectEnabled = () => {
    const { videoProjectQuery } = props;

    // disable delete project button if either there is no project id OR project has been published
    return ( !projectId || videoProjectQuery?.project?.status !== 'DRAFT' );
  };

  const handleExit = () => {
    props.router.push( { pathname: '/admin/dashboard' } );
  };

  const handleDeleteConfirm = async () => {
    const { deleteVideoProject } = props;

    const deletedProjectId = await deleteVideoProject( { variables: { id: projectId } } ).catch(
      err => {
        setError( err );
      },
    );

    if ( deletedProjectId ) {
      handleExit();
    }
  };

  // const handleDeleteCancel = () => {
  //   setDeleteConfirmOpen( false );
  // };

  const handleFinalReview = e => {
    // Prevent multiple clicks - multiple clicks resulted in project going into PUBLISHING status
    if ( !e.detail || e.detail === 1 ) e.target.disabled = true;
    props.router.push(
      {
        pathname: '/admin/project',
        query: {
          content: 'video',
          id: projectId,
        },
      },
      `/admin/project/video/${projectId}/review`,
    );
  };

  const handleSaveDraft = async ( id, projectTitle, tags ) => {
    const {
      updateVideoProject,
      imageUsesQuery: { imageUses },
    } = props;
    const data = buildUpdateVideoProjectTree( filesToUpload, imageUses[0].id, projectTitle, tags );

    await updateVideoProject( {
      variables: {
        data,
        where: {
          id,
        },
      },
    } ).catch( err => console.dir( err ) );
  };

  const handleUploadProgress = async ( progressEvent, file ) => {
    props.uploadProgress( file.id, progressEvent );
  };

  const handleUploadComplete = () => {
    const { uploadReset } = props;

    setIsUploading( false );
    uploadReset(); // reset upload redux store

    updateNotification( 'Project saved as draft' );
    startTimeout();
  };

  const handleUpload = async ( project, tags ) => {
    const { id, projectTitle } = project;
    const { uploadExecute, updateFile } = props;

    // If there are files to upload, upload them
    if ( filesToUpload && filesToUpload.length ) {
      setIsUploading( true );

      // 1. Upload files to S3 & Vimeo and fetch file meta data
      await uploadExecute( id, filesToUpload, handleUploadProgress, updateFile );

      // 2. once all files have been uploaded, create and save new project (only new)
      handleSaveDraft( id, projectTitle, tags );

      // 3. clean up upload process
      handleUploadComplete();

      // 4. set project id to newly created project (what if exsiting project?)
      setProjectId( id );

      // 5. update url to reflect a new project (only new)
      addProjectIdToUrl( id );
    }
  };

  // const renderConfirm = ( isOpen, onConfirm, onCancel ) => (
  //   <Fragment>
  //     <Confirm
  //       className="confirm-modal"
  //       open={ isOpen }
  //       onCancel={ onCancel }
  //       onConfirm={ onConfirm }
  //       cancelButton="No"
  //       confirmButton="Yes"
  //       content={ (
  //         <ConfirmModalContent
  //           className="content"
  //           headline="Unsaved Changes!"
  //         >
  //           <p>You have not finished entering and uploading your project data and content.</p>
  //           <p>Navigating away from this page now, your changes will not be saved and uploads for this project will be canceled.</p>
  //           <p><strong>Do you want to leave this page and lose your changes and cancel your uploads?</strong></p>
  //         </ConfirmModalContent>
  //       ) }
  //     />
  //   </Fragment>
  // );

  const contentStyle = {
    border: `3px solid ${projectId ? 'transparent' : '#02bfe7'}`,
  };

  const { showNotification, notificationMessage } = notification;

  return (
    <div className="edit-project">
      <div className="edit-project__header">
        <ProjectHeader icon="video camera" text="Project Details">
          <ActionButtons
            deleteConfirmOpen={ deleteConfirmOpen }
            setDeleteConfirmOpen={ setDeleteConfirmOpen }
            disabled={ {
              'delete': deleteProjectEnabled(),
              save: !projectId || disableBtns || !isFormValid,
              review: !projectId || disableBtns || !isFormValid,
            } }
            handle={ {
              deleteConfirm: handleDeleteConfirm,
              save: handleExit,
              review: handleFinalReview,
            } }
            show={ {
              'delete': true,
              save: true,
              review: true,
            } }
          />
        </ProjectHeader>
      </div>

      <div style={ centeredStyles }>
        <ApolloError error={ error } />
      </div>

      { /* Form data saved notification */ }
      <Notification
        el="p"
        customStyles={ { ...centeredStyles, top: '8em' } }
        show={ showNotification }
        msg={ notificationMessage }
      />

      { /* Video file notification */ }
      { !isUploading && disableBtns && (
        <Notification
          el="p"
          classes="includeVideoFileNotification"
          customStyles={ uploadVideoFileNotificationStyles }
          msg="Please include a Video file to your project."
        />
      ) }

      { /* upload progress  */ }
      <UploadProgress
        className="edit-project__status alpha"
        projectId={ projectId }
        filesToUpload={ filesToUpload }
        isUploading={ isUploading }
      />

      { /* project details form */ }
      <div className="edit-project__content" style={ contentStyle }>
        <VideoProjectDetailsForm
          id={ projectId }
          updateNotification={ updateNotification }
          handleUpload={ handleUpload }
          maxCategories={ MAX_CATEGORY_COUNT }
          setIsFormValid={ setIsFormValid }
          startTimeout={ startTimeout }
        />
      </div>

      { /* upload progress */ }
      <UploadProgress
        className="edit-project__status beta"
        projectId={ projectId }
        filesToUpload={ filesToUpload }
        isUploading={ isUploading }
      />

      <UploadContext.Provider value={ isUploading }>
        { /* support files */ }
        <div className="edit-project__support-files">
          <VideoProjectSupportFiles projectId={ projectId } heading="Support Files" />
        </div>

        { /* project thumbnails */ }
        <div className="edit-project__items">
          <ProjectUnits
            projectId={ projectId }
            heading="Videos in Project"
            extensions={ ['.mov', '.mp4'] }
          />
        </div>
      </UploadContext.Provider>
    </div>
  );
};

VideoEdit.propTypes = {
  id: PropTypes.string,
  imageUsesQuery: PropTypes.object,
  updateVideoProject: PropTypes.func,
  deleteVideoProject: PropTypes.func,
  videoProjectQuery: PropTypes.object,
  videoProjectUnitsQuery: PropTypes.object,
  router: PropTypes.object,
  setIsUploading: PropTypes.func, // from redux
  uploadExecute: PropTypes.func, // from redux
  uploadReset: PropTypes.func, // from redux
  uploadProgress: PropTypes.func, // from redux
  updateFile: PropTypes.func,
  upload: PropTypes.object, // from redux
};

const mapStateToProps = state => ( {
  upload: state.upload,
} );

const connectUnitThumbnails = async ( props, result ) => {
  const {
    updateVideoProject: { units, thumbnails },
  } = result;
  const { updateVideoUnit } = props;

  if ( units.length && thumbnails.length ) {
    units.forEach( async unit => {
      let thumbnail = thumbnails.find( tn => tn.language.id === unit.language.id );

      // if an applicable language thumbnail doesn't exist, use english version if available
      thumbnail = thumbnail || thumbnails.find( tn => tn.language.locale === 'en-us' );

      if ( thumbnail ) {
        updateVideoUnit( {
          variables: {
            data: buildThumbnailTree( thumbnail ),
            where: {
              id: unit.id,
            },
          },
        } ).catch( err => console.dir( err ) );
      }
    } );
  }
};

const refetchVideoProject = result => {
  // This intermittently throws an error due to a null result
  // null result results in returning a null query

  // rerun sub queries, form, units, supportfiles
  try {
    const { data } = result;
    const node = data.updateVideoProject || data.updateVideoUnit;

    return [
      {
        query: VIDEO_PROJECT_FILES_QUERY,
        variables: { id: node.id },
      },
      {
        query: VIDEO_PROJECT_UNITS_QUERY,
        variables: { id: node.id },
      },
    ];
  } catch ( err ) {
    console.log( err );
  }
};

export default compose(
  withRouter,
  withFileUpload,
  connect( mapStateToProps, actions ),
  graphql( VIDEO_PROJECT_QUERY, {
    // this is not the best way to fetch project status, update after release
    name: 'videoProjectQuery',
    options: props => ( {
      variables: { id: props.id },
    } ),
    skip: props => !props.id,
  } ),
  graphql( IMAGE_USES_QUERY, {
    name: 'imageUsesQuery',
  } ),
  graphql( VIDEO_PROJECT_UNITS_QUERY, {
    name: 'videoProjectUnitsQuery',
    options: props => ( {
      variables: { id: props.id },
    } ),
    skip: props => !props.id,
  } ),
  graphql( DELETE_VIDEO_PROJECT_MUTATION, { name: 'deleteVideoProject' } ),
  graphql( UPDATE_VIDEO_UNIT_MUTATION, {
    name: 'updateVideoUnit',
    options: () => ( {
      refetchQueries: result => refetchVideoProject( result ),
    } ),
  } ),
  graphql( UPDATE_VIDEO_PROJECT_MUTATION, {
    name: 'updateVideoProject',
    options: props => ( {
      refetchQueries: result => refetchVideoProject( result ),
      onCompleted: result => connectUnitThumbnails( props, result ),
    } ),
  } ),
)( VideoEdit );
