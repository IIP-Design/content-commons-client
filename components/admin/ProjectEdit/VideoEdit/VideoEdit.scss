@import "styles/colors.scss";
@import "styles/mixins.scss";

.edit-project {
  @include uppercase;

  @media screen and (max-width: 550px) {
    .project_buttons {
      @include multi-column-layout;
    }
  }

  .includeVideoFileNotification {
    opacity: 1 !important;
    transition: opacity 1.5s ease;
  }

  .ui.button {
    @include base-button;

    &.basic.btn--edit {
      color: $light-blue !important;

      &:active,
      &:hover,
      &:focus {
        color: $matisse-blue !important;
        box-shadow: none;
      }
    }
  }

  @media screen and (min-width: 992px) {
    &__form--save {
      width: 100%;
    }
  }

  &__add-more {
    &.ui.basic.button {
      @include ghost-button;
    }
  }

  &__add-more__wrapper {
    margin-top: 0;
    padding-left: 1.75em;
    @media screen and (max-width: 767px) {
      padding-left: 1rem;
    }
  }

  &__items {
    /**
     * use rem for l & r padding to
     * match and override Semantic UI
     */
    padding: 1.75em 1rem;
  }

  &__content,
  &__support-files {
    padding: 1.75em 0;
  }

  &__content,
  &__support-files,
  &__items {
    @media screen and (min-width: 768px) {
      padding: 1.75em;
    }
  }

  &__support-files,
  &__items {
    border-top: 1px solid $light-grey;
  }

  .ui.modal {
    > .content {
      position: relative;
    }
  }
}
