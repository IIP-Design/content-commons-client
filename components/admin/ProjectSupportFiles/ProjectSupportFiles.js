/**
 *
 * ProjectSupportFiles
 *
 */

import React, { Fragment } from 'react';
import {
  bool, func, object, string
} from 'prop-types';
import { Checkbox, Grid } from 'semantic-ui-react';

import IconPopup from 'components/popups/IconPopup/IconPopup';
import SupportFileTypeList from 'components/admin/projects/ProjectEdit/SupportFileTypeList/SupportFileTypeList';

import './ProjectSupportFiles.scss';

const ProjectSupportFiles = props => {
  const {
    heading,
    projectId,
    supportFiles,
    hasSubmittedData,
    protectImages,
    handleChange,
    config,
    hasUploaded
  } = props;

  if ( Object.keys( supportFiles ).length === 0 ) return null;

  const fileTypes = Object.keys( config );
  const columns = fileTypes.filter( n => supportFiles[n].length > 0 );

  const renderFileTypes = type => {
    const {
      headline,
      popupMsg,
      checkBoxLabel,
      checkBoxName,
      iconMsg,
      iconSize,
      iconType
    } = config[type];

    return (
      <Fragment key={ `file-type-${type}` }>
        <Grid.Column>
          <SupportFileTypeList
            headline={ headline }
            projectId={ projectId }
            fileType={ type }
            popupMsg={ popupMsg }
            data={ supportFiles[type] }
            hasSubmittedData={ hasSubmittedData }
            hasUploaded={ hasUploaded }
          />

          { ( type === 'other' && hasUploaded )
            && (
              <Fragment>
                <Checkbox
                  id="protect-images"
                  label={ (
                    /* eslint-disable jsx-a11y/label-has-for */
                    <label htmlFor="protect-images">
                      { checkBoxLabel }
                    </label>
                  ) }
                  name={ checkBoxName }
                  type="checkbox"
                  checked={ protectImages }
                  onChange={ handleChange }
                />
                <IconPopup
                  message={ iconMsg }
                  iconSize={ iconSize }
                  iconType={ iconType }
                  popupSize="mini"
                />
              </Fragment>
            ) }
        </Grid.Column>
      </Fragment>
    );
  };

  return (
    <div className="support-files">
      <Grid stackable>
        <Grid.Row>
          <Grid.Column width={ 16 }>
            <h3 className="heading uppercase">{ heading }</h3>
          </Grid.Column>
        </Grid.Row>

        <Grid.Row columns={ columns.length } divided>
          { fileTypes.map( renderFileTypes ) }
        </Grid.Row>
      </Grid>
    </div>
  );
};

ProjectSupportFiles.propTypes = {
  heading: string,
  projectId: string.isRequired,
  supportFiles: object.isRequired,
  hasSubmittedData: bool,
  protectImages: bool,
  handleChange: func.isRequired,
  config: object.isRequired,
  hasUploaded: bool
};

export default ProjectSupportFiles;
