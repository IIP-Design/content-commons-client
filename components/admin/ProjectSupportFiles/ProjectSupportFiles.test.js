import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { supportFilesConfig } from 'components/admin/projects/ProjectEdit/VideoEdit/VideoEdit';
import ProjectSupportFiles from './ProjectSupportFiles';

const props = {
  heading: 'Support Files',
  projectId: '123',
  supportFiles: {
    srt: [
      { id: 'qwqw' },
      { id: 'erer' },
      { id: 'asas' },
      { id: 'dfdf' }
    ],
    other: [
      { id: 'tyty' },
      { id: 'ghgh' },
      { id: 'vbvb' }
    ]
  },
  hasSubmittedData: false,
  protectImages: true,
  handleChange: jest.fn(),
  config: supportFilesConfig,
  hasUploaded: true
};

const Component = <ProjectSupportFiles { ...props } />;

describe( '<ProjectSupportFiles />', () => {
  it( 'renders without crashing', () => {
    const wrapper = shallow( Component );
    expect( wrapper.exists() ).toEqual( true );
    expect( toJSON( wrapper ) ).toMatchSnapshot();
  } );

  it( 'renders the heading', () => {
    const wrapper = shallow( Component );
    const heading = wrapper.find( '.heading' );
    expect( heading.name() ).toEqual( 'h3' );
    expect( heading.text() ).toEqual( props.heading );
  } );

  it( 'renders the correct number of `SupportFileTypeList`s', () => {
    const wrapper = shallow( Component );
    const lists = wrapper.find( 'SupportFileTypeList' );
    const typesCount = Object.keys( props.config ).length;
    expect( lists ).toHaveLength( typesCount );
  } );

  it( 'renders a `Checkbox` for protect images if the type is `other`', () => {
    const wrapper = shallow( Component );

    expect( wrapper.find( 'Checkbox' ).exists() )
      .toEqual( true );
    wrapper.setProps( {
      config: {
        srt: {
          headline: 'Test Files',
          fileType: 'abc',
          popupMsg: 'A test message'
        }
      }
    } );
    expect( wrapper.find( 'Checkbox' ).exists() )
      .toEqual( false );
  } );

  it( 'checking/unchecking `Checkbox` calls `handleChange`', () => {
    const wrapper = shallow( Component );
    const checkbox = wrapper.find( 'Checkbox' );
    if ( checkbox.exists() ) {
      checkbox.simulate( 'change' );
      expect( props.handleChange ).toHaveBeenCalled();
    }
  } );

  it( 'does not render if no support files', () => {
    const wrapper = shallow( Component );
    wrapper.setProps( { supportFiles: {} } );
    expect( wrapper.find( '.support-files' ) )
      .toHaveLength( 0 );
  } );
} );
