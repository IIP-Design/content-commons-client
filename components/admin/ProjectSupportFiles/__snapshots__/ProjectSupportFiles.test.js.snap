// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectSupportFiles /> renders without crashing 1`] = `
<div
  className="support-files"
>
  <Grid
    stackable={true}
  >
    <GridRow>
      <GridColumn
        width={16}
      >
        <h3
          className="heading uppercase"
        >
          Support Files
        </h3>
      </GridColumn>
    </GridRow>
    <GridRow
      columns={2}
      divided={true}
    >
      <GridColumn>
        <SupportFileTypeList
          data={
            Array [
              Object {
                "id": "qwqw",
              },
              Object {
                "id": "erer",
              },
              Object {
                "id": "asas",
              },
              Object {
                "id": "dfdf",
              },
            ]
          }
          fileType="srt"
          hasSubmittedData={false}
          hasUploaded={true}
          headline="SRT Files"
          popupMsg="Some info about what SRT files are."
          projectId="123"
        />
      </GridColumn>
      <GridColumn>
        <SupportFileTypeList
          data={
            Array [
              Object {
                "id": "tyty",
              },
              Object {
                "id": "ghgh",
              },
              Object {
                "id": "vbvb",
              },
            ]
          }
          fileType="other"
          hasSubmittedData={false}
          hasUploaded={true}
          headline="Additional Files"
          popupMsg="Additional files that can be used with this video, e.g., audio file, pdf."
          projectId="123"
        />
        <Checkbox
          checked={true}
          id="protect-images"
          label={
            <label
              htmlFor="protect-images"
            >
              Disable right-click to protect your images
            </label>
          }
          name="protectImages"
          onChange={[MockFunction]}
          type="checkbox"
        />
        <IconPopup
          iconSize="small"
          iconType="info circle"
          message="Checking this prevents people from downloading and using your images. Useful if your images are licensed."
          popupSize="mini"
        />
      </GridColumn>
    </GridRow>
  </Grid>
</div>
`;
